manta-trusted-setup/src/groth16/ceremony/signature.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Groth16 Trusted Setup Ceremony Signatures
+
+use alloc::vec::Vec;
+use manta_crypto::signature;
+use manta_util::{serde::Serialize, AsBytes};
+
+/// Nonce
+pub trait Nonce: PartialEq {
+    /// Increments the current nonce by one.
+    fn increment(&self) -> Self;
+
+    /// Checks if the current nonce is valid.
+    fn is_valid(&self) -> bool;
+}
+
+impl Nonce for u64 {
+    #[inline]
+    fn increment(&self) -> Self {
+        self.saturating_add(1)
+    }
+
+    #[inline]
+    fn is_valid(&self) -> bool {
+        *self != Self::MAX
+    }
+}
+
+/// Checks if the two nonces, `current` and `user_nonce`, are both valid and equal to each other.
+#[inline]
+pub fn check_nonce<N>(current: &N, user_nonce: &N) -> bool
+where
+    N: Nonce,
+{
+    current.is_valid() && user_nonce.is_valid() && current == user_nonce
+}
+
+/// Message with Nonce
+#[derive(Clone, Debug, Default, Eq, Hash, PartialEq)]
+pub struct Message<N> {
+    /// Nonce
+    pub nonce: N,
+
+    /// Encoded Message
+    pub encoded_message: Vec<u8>,
+}
+
+impl<N> AsBytes for Message<N>
```

Comment:
`AsBytes` has a single function that converts from `&self` to `Vec<u8>` whereas `serde::Serialize` is an entire framework for transforming types into any possible serialization system. For the `ed25519` signature we require at least `AsBytes` since they actually ask for `&[u8]` in our interface and that's too restrictive, we loosen the requirement to `AsBytes`.